### DISCUSSION TITLE
How to fix the crash issue with exception message "Indirect call guard check detected invalid control" in openssl 3.0.14?

### DISCUSSION THREAD

**Discussion by Bo-Liu-BL** [openssl/openssl#25342] [Q&A] [answered]
Hi experts,

I encountered a strange issue:

- When using the openssl **3.0.13**, my program worked fine.
- But upgrading to openssl **3.0.14**, there was always crash.

The exception message was:
`Unhandled exception at 0x00007FFF5EC75180 (ntdll.dll) in My_Private_Program.exe: Indirect call guard check detected invalid control transfer.`
Not sure what happened. Google didn't help.

The source code of the function `Poly1305_Update` :
https://github.com/openssl/openssl/blob/master/crypto/poly1305/poly1305.c#L468

![image](https://github.com/user-attachments/assets/f935449e-bdd5-4de5-81cc-c99a493b2d11)

The dump showed that the variable `rem` is optimized away and not available. 
But `num` is 5, `POLY1305_BLOCK_SIZE` is 16, thus `inp` should be stepped forward `16-5=11` bytes.

The info of `inp` in the stack:
`inp | 0x00007ffec5c4f0b0 "" | const unsigned char *`

Looks it contains zero memory:
```
0x00007FFEC5C4F0B0  00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00  .................................
0x00007FFEC5C4F0D1  00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00  .................................
0x00007FFEC5C4F0F2  00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00  .................................
```

Som auto variables are shown as:
![image](https://github.com/user-attachments/assets/93159731-78ff-493f-be1c-f9724c019029)

**Not sure why it crashed with the exception "Indirect call guard check detected invalid control".
Any suggestion is appreciate. Thank you!**

The crash stack trace:
```
ntdll.dll!RtlFailFast2()
ntdll.dll!RtlpHandleInvalidUserCallTarget()
ntdll.dll!LdrpHandleInvalidUserCallTarget()
libcrypto-3-x64.dll!Poly1305_Update(poly1305_context * ctx, const unsigned char * inp, unsigned __int64 len)
libcrypto-3-x64.dll!chacha20_poly1305_aead_cipher(prov_cipher_ctx_st * bctx, unsigned char * out, unsigned __int64 * outl, const unsigned char * in, unsigned __int64 inl)
libcrypto-3-x64.dll!chacha20_poly1305_cipher(void * vctx, unsigned char * out, unsigned __int64 * outl, unsigned __int64 outsize, const unsigned char * in, unsigned __int64 inl)
libcrypto-3-x64.dll!EVP_DecryptUpdate(evp_cipher_ctx_st * ctx, unsigned char * out, int * outl, const unsigned char * in, int inl)
libssl-3-x64.dll!tls13_enc(ssl_st * s, ssl3_record_st * recs, unsigned __int64 n_recs, int sending, ssl_mac_buf_st * mac, unsigned __int64 macsize)
libssl-3-x64.dll!ssl3_get_record(ssl_st * s)
libssl-3-x64.dll!ssl3_read_bytes(ssl_st * s, int type, int * recvd_type, unsigned char * buf, unsigned __int64 len, int peek, unsigned __int64 * readbytes)
libssl-3-x64.dll!tls_get_message_header(ssl_st * s, int * mt)
libssl-3-x64.dll!read_state_machine(ssl_st * s)
libssl-3-x64.dll!state_machine(ssl_st * s, int server)
libssl-3-x64.dll!SSL_do_handshake(ssl_st * s)
Xxx_My_Private_Library.dll!InterCalc_SslHandshake(CbData *data)
```

**Comment by nhorman** [comment, marked as answer]
I assume you are building with /guard:cf enabled here?  The problem isn't with the variable you are looking at, its with the the call to poly1305_blocks which expects a void * as the first argument, but recieives a double * instead.  It works, because most compilers treat the pointers as equivalent, but strictly speaking its an undefined behavior, and so the /guard:cf argument flags it.  There are many such cases in openssl, and while older versions of clang/vs haven't flagged this, its started to happen more and more as compilers enable this check.

**Reply by paulidale** [reply]
Just noting that this isn't undefined behaviour: the C standard allows conversion of any pointer to void *.  That doesn't mean a compiler won't warn about it, casting a double * to a void * is often done by mistake.

**Comment by Bo-Liu-BL** [comment]
Thank you for taking care of the discussion, @nhorman and @paulidale .

Yeah, `/guard:cf` is used when building the product.

**Reply by nhorman** [reply]
You're going to encounter hundreds of these errors when building with /guard:cf enabled.  The only real choice at the moment is to build with it disabled at the moment to avoid the run time failures.

That said patches to fix up the calling conventions as you find them would be welcome.
